project-1/main.py

```diff
@@ -0,0 +1,193 @@
+# pylint: disable=import-error
+from pathlib import Path
+from google.oauth2 import service_account
+from gspread_dataframe import set_with_dataframe
+from exceptions import *
+import gspread
+import pandas as pd
+import logging
+import subprocess
+
+
+# logging config
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[logging.StreamHandler()],
+)
+
+CREDENTIALS_PATH = "./credentials.json"
+BASH_SCRIPT_PATH = "./extract.sh"
+
+
+class Workbook:
+    #  initialize gspread connections
+    def __init__(
+        self, workbook_name: str, sheet_name: str, email: str, year: str, month: str
+    ) -> None:
+        self.workbook_name = workbook_name
+        self.sheet_name = sheet_name
+        self.email = email
+        self.year = year
+        self.month = month
+
+        # avoid attribute-defined-outside-init
+        self.sh = None
+        self.worksheet = None
+        self.processed_data = None
+        self.file_path = None
+        self.df = None
+
+        # connection parameters
+        scope = [
+            "https://spreadsheets.google.com/feeds",
+            "https://www.googleapis.com/auth/spreadsheets",
+            "https://www.googleapis.com/auth/drive.file",
+            "https://www.googleapis.com/auth/drive",
+        ]
+
+        try:
+            # google cloud platform API credentials
+            creds = service_account.Credentials.from_service_account_file(
+                CREDENTIALS_PATH, scopes=scope
+            )
+            # create API client
+            self.client = gspread.authorize(creds)
+
+            logging.info("Connected to Google Sheets successfully.")
+        except Exception as e:
+            raise NewSpreadsheetError("Could not connect to Google Sheets") from e
+
+    def _run_bash_process(self) -> Path:
+        """_summary_
+            utility function to execute bash script using a subprocess to download data
+        Returns:
+            Path: returns downloaded file path
+        """
+
+        # bash script name
+        bash_script = BASH_SCRIPT_PATH
+
+        # bash positional arguments
+        arguments = [f"{self.year}", f"{self.month}"]
+
+        # run bash as script as subprocess
+        result = subprocess.run(
+            ["bash", bash_script, *arguments],
+            check=True,
+            text=True,
+            stdout=subprocess.PIPE,
+        )
+        script_output = result.stdout
+        download_path = script_output.split("\n")[-2]
```

Comment:
Can you comment an example for this string extraction, like what is the value we're trying to extract and what does what we finally extract look like